Add a limits config option to allow for dropping of the cluster label

cstyan opened the pull request:
Fixes #1724 

Adds an optional flag per user to drop the cluster label, in addition to dropping the replica label, when ingesting HA samples. Should only be used if that user has their own set of more than two labels that they use to uniquely identify Prometheus replicas.

There's a bit of a refactor here as well:
- I noticed that in the config the ordering of the values was cluster label then replica label, but in RegisterFlags the order is replica then cluster. I moved them around in RegisterFlags to match the order in the struct.
- `distributor.Push` was getting pretty long, I moved the internals of the validation loop into a function `validateSeries`. This also had the side effect of making that portion of the code testable. I've only added tests for the label removal bits. I assume the actual validation functions `ValidateLabels` and `ValidateSamples` are tested already.
- renamed removeReplicaLabel -> removeLabel, it wasn't doing anything replica label specific anyways

Signed-off-by: Callum Styan <callumstyan@gmail.com>
cc @tomwilkie @gouthamve

bboreham commented on docs/arguments.md:
```diff
@@ -144,11 +144,13 @@ prefix these flags with `distributor.ha-tracker.`
 ### HA Tracker
 
-HA tracking has two of it's own flags:
+HA tracking has three of it's own flags:
```
```suggestion
HA tracking has three of its own flags:
```

bboreham commented:
I find it easier to review PRs like this when you split the pure-refactoring part from the behaviour-changing part, into separate commits. Otherwise I have to figure out your intention on every line and that takes longer.

bboreham commented on pkg/distributor/distributor.go:
```diff
@@ -280,14 +281,55 @@ func (d *Distributor) checkSample(ctx context.Context, userID, cluster, replica
 	return true, nil
 }
 
+// Validates a single series from a write request. Will remove HA labels if necessary.
```
State what the return values are.
Include why we have return values both via pointer and by value.

cstyan replied:
Done. I've simplified the function as well so the pointer isn't needed anymore.

bboreham commented on pkg/util/validation/limits.go:
```diff
@@ -48,8 +49,9 @@ func (l *Limits) RegisterFlags(f *flag.FlagSet) {
-	f.StringVar(&l.HAReplicaLabel, "distributor.ha-tracker.replica", "__replica__", "Prometheus label to look for in samples to identify a Prometheus HA replica.")
 	f.StringVar(&l.HAClusterLabel, "distributor.ha-tracker.cluster", "cluster", "Prometheus label to look for in samples to identify a Prometheus HA cluster.")
+	f.StringVar(&l.HAReplicaLabel, "distributor.ha-tracker.replica", "__replica__", "Prometheus label to look for in samples to identify a Prometheus HA replica.")
+	f.BoolVar(&l.HADropClusterLabel, "distributor.ha-tracker.drop-cluster-label", false, "Enable this flag to drop the cluster label when ingesting HA samples for a user, in addition to dropping the replica label. Note that this flag should only be used when users already have their own combination of more than two labels to uniquely identify Prometheus replicas.")
```
The 'when' part doesn't appear to be true - it is done unconditionally.

cstyan replied:
fixed

cstyan commented on pkg/distributor/distributor.go:
```diff
@@ -249,11 +250,11 @@ func shardByAllLabels(userID string, labels []client.LabelAdapter) (uint32, erro
-// Remove the replica label from a slice of LabelPairs if it exists.
-func removeReplicaLabel(replica string, labels *[]client.LabelAdapter) {
+// Remove the label labelname from a slice of LabelPairs if it exists.
+func removeLabel(labelName string, labels *[]client.LabelAdapter) {
```
I'll have a look at the labelsPairs function in the morning.

cstyan replied:
`labelPairs` is nested within the `ingester` package. If we're going to merge the two functions should that exist in a new package under `util`?

And the reason for preserving order is only that I've generally seen that be the case within Prometheus itself. For example within remote write, when processing a sample through external labelse and relabel rules, we preserve the sorted order of the labels.

cstyan commented:
@bboreham thanks for the review! I've split out the refactoring into another commit. I also changed the `validateSeries` function to no longer require the error pointer argument in addition to returning an error.

bboreham commented:
Given that it isn't tied to HA, could the feature be done as a general label-removal mechanism? This would give us a portion of #771.

cstyan commented:
I moved the cluster label dropping inside the check for `removeReplica`, which is tied to HA. I can move that back out and turn the label removal into a more general non-HA dependent thing, maybe in another PR? if you'd like.

bboreham commented:
Given that it doesn't cost me anything, yes I would prefer a generic "for tenant X, everywhere you see label Y, drop it".

nbellowe commented:
@cstyan Any updates with this PR? Its much appreciated, we're waiting on this to enable HA writing to cortex.

bboreham commented on docs/arguments.md:
```diff
@@ -144,11 +144,13 @@ prefix these flags with `distributor.ha-tracker.`
    Prometheus label to look for in samples to identify a Prometheus HA cluster. (default "cluster")
 - `distributor.ha-tracker.replica`
    Prometheus label to look for in samples to identify a Prometheus HA replica. (default "`__replica__`")
+- `distributor.ha-tracker.drop-cluster-label`
```
does this name need to be updated?

cstyan replied:
Yes, thanks for catching that.

bboreham commented on pkg/util/validation/limits.go:
```diff
@@ -48,8 +51,9 @@ func (l *Limits) RegisterFlags(f *flag.FlagSet) {
 	f.BoolVar(&l.AcceptHASamples, "distributor.ha-tracker.enable-for-all-users", false, "Flag to enable, for all users, handling of samples with external labels identifying replicas in an HA Prometheus setup.")
-	f.StringVar(&l.HAReplicaLabel, "distributor.ha-tracker.replica", "__replica__", "Prometheus label to look for in samples to identify a Prometheus HA replica.")
 	f.StringVar(&l.HAClusterLabel, "distributor.ha-tracker.cluster", "cluster", "Prometheus label to look for in samples to identify a Prometheus HA cluster.")
+	f.StringVar(&l.HAReplicaLabel, "distributor.ha-tracker.replica", "__replica__", "Prometheus label to look for in samples to identify a Prometheus HA replica.")
```
It's not obvious to me why this moved.  No big deal.

cstyan replied:
Sorry for the noise. I just rearranged the calls to `flags.VarType` here for the cluster and replica labels to match the order those values are specified in the Limits struct.

gouthamve commented on pkg/distributor/distributor.go:
```diff
@@ -280,6 +281,48 @@ func (d *Distributor) checkSample(ctx context.Context, userID, cluster, replica
 }
 
+// Validates a single series from a write request. Will remove HA labels if necessary.
+// Takes a pointer for a partial error so that we can get partial errors, errors during validation
```
Can you update the comment to remove the reference to the pointer?

gouthamve commented on pkg/util/validation/limits.go:
```diff
@@ -142,14 +146,19 @@ func (o *Overrides) AcceptHASamples(userID string) bool {
-// HAClusterLabel returns the cluster label to look for when deciding whether to accept a sample from a Prometheus HA replica.
-func (o *Overrides) HAClusterLabel(userID string) string {
-	return o.overridesManager.GetLimits(userID).(*Limits).HAClusterLabel
+// DropLabels returns whether the cluster label should be dropped when ingesting HA samples for the user.
```
Returns the list of labels to be dropped?

gouthamve commented on pkg/distributor/distributor.go:
```diff
@@ -280,6 +281,48 @@ func (d *Distributor) checkSample(ctx context.Context, userID, cluster, replica
+	if removeReplica {
+		removeLabel(d.limits.HAReplicaLabel(userID), &ts.Labels)
+	}
+	for _, s := range d.limits.DropLabels(userID) {
```
`s` is a weird variable name in this instance, maybe call it `label`/`labelName`?

gouthamve commented on pkg/distributor/distributor.go:
```diff
@@ -280,6 +281,48 @@ func (d *Distributor) checkSample(ctx context.Context, userID, cluster, replica
+	for _, s := range d.limits.DropLabels(userID) {
+		removeLabel(s, &ts.Labels)
+	}
+	key, err := d.tokenForLabels(userID, ts.Labels)
```
Also check and handle if the series has no labels anymore after the removing of labels.

nbellowe commented:
Thank you @cstyan!
